crates/ruff_server/src/trace.rs

```diff
@@ -0,0 +1,143 @@
+            .with_timer(Uptime::default())
+            .with_thread_names(true)
+            .with_ansi(false)
+            .with_writer(TracingWriter::new(log_file))
+            .with_filter(TraceLevelFilter)
+            .with_filter(LogLevelFilter { filter: log_level }),
+    );
+
+    tracing::subscriber::set_global_default(subscriber)
+        .expect("should be able to set global default subscriber");
+}
+
+#[derive(Clone, Copy, Debug, Deserialize, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(rename_all = "lowercase")]
+pub(crate) enum LogLevel {
+    #[default]
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl LogLevel {
+    fn trace_level(self) -> tracing::Level {
+        match self {
+            Self::Error => tracing::Level::ERROR,
+            Self::Warn => tracing::Level::WARN,
+            Self::Info => tracing::Level::INFO,
+            Self::Debug => tracing::Level::DEBUG,
+            Self::Trace => tracing::Level::TRACE,
+        }
+    }
+}
+
+struct LogLevelFilter {
+    filter: LogLevel,
+}
+
+struct TraceLevelFilter;
+
+impl<S> tracing_subscriber::layer::Filter<S> for LogLevelFilter {
+    fn enabled(
+        &self,
+        meta: &tracing::Metadata<'_>,
+        _: &tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        let filter = if meta.target().starts_with("ruff") {
+            self.filter.trace_level()
+        } else {
+            tracing::Level::INFO
+        };
+
+        meta.level() <= &filter
+    }
+
+    fn max_level_hint(&self) -> Option<tracing::level_filters::LevelFilter> {
+        Some(LevelFilter::from_level(self.filter.trace_level()))
+    }
+}
+
+impl<S> tracing_subscriber::layer::Filter<S> for TraceLevelFilter {
+    fn enabled(
+        &self,
+        _: &tracing::Metadata<'_>,
+        _: &tracing_subscriber::layer::Context<'_, S>,
+    ) -> bool {
+        trace_value() != lsp_types::TraceValue::Off
+    }
+}
+
+struct TracingWriter(Option<PathBuf>);
+
+impl TracingWriter {
+    fn new(file: Option<&std::path::Path>) -> TracingWriter {
+        Self(file.map(std::path::Path::to_path_buf))
+    }
+}
+
+impl tracing_subscriber::fmt::MakeWriter<'_> for TracingWriter {
+    type Writer = Box<dyn std::io::Write>;
+
+    fn make_writer(&self) -> Self::Writer {
+        if let Some(file) = self.0.as_ref().and_then(|path| {
+            std::fs::OpenOptions::new()
+                .create(true)
+                .append(true)
+                .open(path)
+                .ok()
+        }) {
+            Box::new(file)
+        } else {
+            Box::new(std::io::stderr())
+        }
+    }
+}
+
+#[inline]
+fn trace_value() -> lsp_types::TraceValue {
+    std::env::var(TRACE_ENV_KEY)
+        .ok()
+        .and_then(|trace| serde_json::from_value(serde_json::Value::String(trace)).ok())
+        .unwrap_or_else(|| {
+            *TRACE_VALUE
+                .lock()
+                .expect("trace value mutex should be available")
+        })
+}
```

Comment:
I would find some module level documentation helpful that explains the different ways on how tracing can be configured. It seem we now support an env var, `TRACE_VALUE`, a file, and the PR description mentions a `logLevel` setting, but it's not clear to me where that one is set.